### PR TITLE
Initial purge private data integration tests

### PR DESCRIPTION
**Pull request by jt-nti**
Work in progress to add private data purge integration tests

#### Type of change

- Test update

#### Related issues

#3030 

[Private data purge RFC](https://github.com/hyperledger/fabric-rfcs/blob/main/text/0000-private_data_purge.md)

Signed-off-by: James Taylor <jamest@uk.ibm.com>

### CODE REVIEW

**Comment by yacovm** [comment]
Let's redirect this PR to release-2.5 branch

**Comment by jt-nti** [comment]
I don't see any reason to do that. I would prefer to continue to deliver PRs to main and cherry pick commits back to any previous releases which they're relevant for.

File: integration/pvtdata/data_purge_test.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"syscall"
+
+	"github.com/hyperledger/fabric/integration/nwo"
+	"github.com/hyperledger/fabric/integration/pvtdata/marblechaincodeutil"
+	. "github.com/onsi/ginkgo"
```
Comment (denyeart) [inline integration/pvtdata/data_purge_test.go]:
```suggestion
	. "github.com/onsi/ginkgo/v2"
```

File: go.mod [hunk 2]
```diff
@@ -48,7 +48,10 @@ require (
 	gopkg.in/yaml.v2 v2.4.0
 )
 
-require google.golang.org/protobuf v1.27.1
+require (
+	github.com/onsi/ginkgo v1.16.4
```
Comment (denyeart) [inline go.mod]: github.com/onsi/ginkgo/v2 is already added in the above require. No need to add a second (older) instance of ginkgo.

**Comment by denyeart** [comment]
I don't think there is any reason to change go.mod or vendor directory in this PR.

File: go.mod [hunk 3]
```diff
@@ -48,7 +48,10 @@ require (
...
+	google.golang.org/protobuf v1.27.1
```
Comment (denyeart) [inline go.mod]: Also, why was a second protobuf recently added in its own require block recently? If the intent is to update from
`github.com/golang/protobuf v1.5.2`
to
`google.golang.org/protobuf v1.27.1`
shouldn't that be done for the entire repo rather than add a second protobuf?

Comment (denyeart) [reply go.mod]: @jt-nti I still have this question about the change in https://github.com/hyperledger/fabric/pull/3522

File: integration/chaincode/marbles_private/chaincode.go [hunk 4]
```diff
@@ -377,6 +380,84 @@ func (t *MarblesPrivateChaincode) delete(stub shim.ChaincodeStubInterface, args
 	return shim.Success(nil)
 }
 
+// =====================================================
+// purge - remove a marble key/value pair from state and
+// remove all trace of private details
+// =====================================================
+func (t *MarblesPrivateChaincode) purge(stub shim.ChaincodeStubInterface, args []string) pb.Response {
+	fmt.Println("- start delete marble")
```
Comment (manish-sethi) [inline integration/chaincode/marbles_private/chaincode.go]: start purge marble?

File: integration/chaincode/marbles_private/chaincode.go [hunk 5]
```diff
@@ -377,6 +380,84 @@ func (t *MarblesPrivateChaincode) delete(stub shim.ChaincodeStubInterface, args
...
+	}
+
+	var marbleToDelete marble
+	err = json.Unmarshal([]byte(valAsbytes), &marbleToDelete)
+	if err != nil {
+		return shim.Error("Failed to decode JSON of: " + string(valAsbytes))
+	}
+
+	// delete the marble from state
+	err = stub.DelPrivateData("collectionMarbles", marblePurgeInput.Name)
```
Comment (manish-sethi) [inline integration/chaincode/marbles_private/chaincode.go]: stub.PurgePrivateData?

File: integration/chaincode/marbles_private/chaincode.go [hunk 6]
```diff
@@ -377,6 +380,84 @@ func (t *MarblesPrivateChaincode) delete(stub shim.ChaincodeStubInterface, args
...
+		return shim.Error("Failed to delete state:" + err.Error())
+	}
+
+	// Also delete the marble from the color~name index
+	indexName := "color~name"
+	colorNameIndexKey, err := stub.CreateCompositeKey(indexName, []string{marbleToDelete.Color, marbleToDelete.Name})
+	if err != nil {
+		return shim.Error(err.Error())
+	}
+	err = stub.DelPrivateData("collectionMarbles", colorNameIndexKey)
```
Comment (manish-sethi) [inline integration/chaincode/marbles_private/chaincode.go]: stub.PurgePrivateData?

File: integration/pvtdata/data_purge_test.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+	})
+
+	// 1. User is able to submit a purge transaction that involves more than one keys
+	PIt("should accept multiple keys for purging in the same transaction")
+
+	// 2. The endorsement policy is evaluated correctly for a purge transaction under different endorsement policy settings (e.g., collection level/ key-hash based)
+	PIt("should correctly enforce collection level endorsement policies")
+	PIt("should correctly enforce key-hash based endorsement policies")
+	PIt("should correctly enforce other endorsement policies (TBC)")
+
```
Comment (manish-sethi) [inline integration/pvtdata/data_purge_test.go:129]: Just to take a note - The endorsement policy level tests need not to be prioritized over other behaviour, and if you prefer, they need not to be very exhaustive because we have not introduced anything new that would change this evaluation as it would have done for any other existing write/delete operations.

File: integration/pvtdata/data_purge_test.go [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+	//    - Submit a collection config update to remove an org
+	//    - Issue a purge transaction to delete few keys
+	//    - The removed orgs peer should have purged the historical versions of intended key
+	PIt("should remove all purged data from a previously eligible peer")
+
+	// 5. A new peer able to reconcile from a purged peer
+	//    - Stop one of the peers of an eligible org
+	//    - Add a few keys into a collection
+	//    - Issue a purge transaction for some of the keys
+	//    - Start the stopped peer and the peer should reconcile the partial available data
```
Comment (manish-sethi) [inline integration/pvtdata/data_purge_test.go:148]: You can add one more test - any further writes to private data after the purge operation are not purged, as a side-effect of previous purge operation.

**Review by manish-sethi — COMMENTED** [review]
Looks good to me. Added a few clarification comments. You should be able to proceed with all the tests described here, except the last one related to reconciliation - [once this PR](https://github.com/hyperledger/fabric/pull/3705) is merged

File: integration/chaincode/marbles_private/chaincode.go (hunk 5 above)
Comment (jt-nti) [reply integration/chaincode/marbles_private/chaincode.go]: I think I had intended to do a delete, followed by a purge, since that seems likely to be the most common scenario.

File: integration/chaincode/marbles_private/chaincode.go (hunk 6 above)
Comment (jt-nti) [reply integration/chaincode/marbles_private/chaincode.go]: I think I had intended to do a delete, followed by a purge, since that seems likely to be the most common scenario.

File: integration/chaincode/marbles_private/chaincode.go (hunk 5 above)
Comment (manish-sethi) [reply integration/chaincode/marbles_private/chaincode.go]: Not sure how deleting from one collection and purging from another collection is going to be a common scenario. Looking at the name of the function `(t *MarblesPrivateChaincode) purge` and given the patterns of other functions in this marbles chaincode, one would expect purging the marble. I don't mind either way as it's an integration test only. But, change the function name in that case.

File: integration/chaincode/marbles_private/chaincode.go (hunk 6 above)
Comment (manish-sethi) [reply integration/chaincode/marbles_private/chaincode.go]: Same as above

File: integration/chaincode/marbles_private/chaincode.go (hunk 5 above)
Comment (jt-nti) [reply integration/chaincode/marbles_private/chaincode.go]: Doh, I wasn't intending to delete from one collection and purge from another.

File: integration/chaincode/marbles_private/chaincode.go [hunk 9]
```diff
@@ -377,6 +380,85 @@ func (t *MarblesPrivateChaincode) delete(stub shim.ChaincodeStubInterface, args
...
+	err = json.Unmarshal([]byte(valAsbytes), &marbleToDelete)
+	if err != nil {
+		return shim.Error("Failed to decode JSON of: " + string(valAsbytes))
+	}
+
+	// delete the marble from state
+	err = stub.PurgePrivateData("collectionMarbles", marblePurgeInput.Name)
+	if err != nil {
+		return shim.Error("Failed to delete state:" + err.Error())
+	}
```
Comment (manish-sethi) [inline integration/chaincode/marbles_private/chaincode.go]: Stale comment. On a side note, the obvious comments does not add much value :-)

File: integration/chaincode/marbles_private/chaincode.go [hunk 10]
```diff
@@ -377,6 +380,85 @@ func (t *MarblesPrivateChaincode) delete(stub shim.ChaincodeStubInterface, args
...
+	// Also delete the marble from the color~name index
+	indexName := "color~name"
+	colorNameIndexKey, err := stub.CreateCompositeKey(indexName, []string{marbleToDelete.Color, marbleToDelete.Name})
+	if err != nil {
+		return shim.Error(err.Error())
+	}
+	err = stub.PurgePrivateData("collectionMarbles", colorNameIndexKey)
+	if err != nil {
+		return shim.Error("Failed to delete state:" + err.Error())
+	}
```
Comment (manish-sethi) [inline integration/chaincode/marbles_private/chaincode.go]: Stale comment.

**Review by manish-sethi — COMMENTED** [review]
A couple clean up comment.

File: integration/pvtdata/data_purge_test.go [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+	PIt("should correctly enforce key-hash based endorsement policies")
+	PIt("should correctly enforce other endorsement policies (TBC)")
+
+	// 3. Data is purged on an eligible peer
+	//    - Add a few keys into a collection
+	//    - Issue a purge transaction for some of the keys
+	//    - Verify that all the versions of the intended keys are purged while the remaining keys still exist
+	//    - Repeat above to purge all keys to test the corner case
+	PIt("should remove all purged data from an eligible peer")
+
```
Comment (manish-sethi) [inline integration/pvtdata/data_purge_test.go:136]: This is a duplicate test of the test at line 106.

File: integration/pvtdata/data_purge_test.go [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+	//    - [Verify that the purge fails]
+	PIt("should fail with an error if the purge capability has not been enabled on the channel")
+
+	It("should prevent purged data being included in responses after the purge transaction has been committed", func() {
+		marblechaincodeutil.AssertPresentInCollectionM(network, channelID, chaincode.Name, `test-marble-0`, org2Peer0)
+		marblechaincodeutil.AssertPresentInCollectionMPD(network, channelID, chaincode.Name, "test-marble-0", org2Peer0)
+
+		marblechaincodeutil.PurgeMarble(network, orderer, channelID, chaincode.Name, `{"name":"test-marble-0"}`, org2Peer0)
+
+		marblechaincodeutil.AssertDoesNotExistInCollectionM(network, channelID, chaincode.Name, `test-marble-0`, org2Peer0)
```
Comment (manish-sethi) [inline integration/pvtdata/data_purge_test.go:112]: This assertion just checks the presence of marble in the private state. In addition, you would need to add the code for testing that the data is not returned in a query for retrieving BlockWithPrivatedata (which retrieves the private data from private data store.)

File: integration/pvtdata/data_purge_test.go (hunk 11 above)
Comment (jt-nti) [reply integration/pvtdata/data_purge_test.go:136]: This test is intended to check that the data is actually gone from the peer, rather than the peer still has the data but is not returning it.

File: integration/pvtdata/data_purge_test.go (hunk 12 above)
Comment (jt-nti) [reply integration/pvtdata/data_purge_test.go:112]: I had a chat with @mbwhite and added a placeholder to make sure purged private data is not included in block event listening replays after the transaction has been committed. Hopefully this PR is in a reasonable state for merging now.

File: integration/pvtdata/data_purge_test.go (hunk 11 above)
Comment (manish-sethi) [reply integration/pvtdata/data_purge_test.go:136]: Sure, we will have to figure out a way to do that in the integration test though! But, I am fine with merging having this for now.

**Review by manish-sethi — APPROVED** [review]

**Comment by manish-sethi** [comment]
@Mergifyio backport release-2.5
